src/metrics/basic.py

```diff
@@ -0,0 +1,281 @@
+    userWhereClause = '({})'.format(' OR '.join(userWhereClauseArray)) if len(userWhereClauseArray) > 0 else None
+    return userWhereClause
+
+def getUserWhereClauseForClickhouse(config):
+    def process(l):
+        data = getGitHubData([l])
+        data = EasyDict(data)
+        if len(data.githubUsers) > 0: return 'actor_id IN [{}]'.format(','.join(data.githubUsers))
+        return None
+    userWhereClauseArray = []
+    if config.get('userIds'): userWhereClauseArray.append('actor_id IN [{}]'.format(','.join(config.get('userIds'))))
+    if config.get('userLogins'): userWhereClauseArray.append('actor_login IN [{}]'.format(list(map(lambda n:'\'{}\''.format(n), config.get('userLogins')))))
+    if config.get('labelIntersect'):
+        return '(' + ' AND '.join(list(filter(lambda i: i != None, list(map(process, config.get('labelIntersect')))))) + ')'
+    if config.get('labelUnion'):
+        data = EasyDict(getGitHubData(config.get('labelUnion')))
+        if len(data.githubRepos > 0): userWhereClauseArray.append('actor_id IN [{}]'.format(','.join(data.githubUsers)))
+    userWhereClause = '({})'.format(' OR '.join(userWhereClauseArray)) if len(userWhereClauseArray) > 0 else None
+    return userWhereClause
+
+# Time
+def getTimeRangeWhereClauseForNeo4j(config, type):
+    timeWhereClauseArray = []
+    forEveryMonthByConfig(config, lambda y, m: timeWhereClauseArray.append('{}.activity_{}{} > 0'.format(type, y, m)))
+    if len(timeWhereClauseArray) == 0: raise Exception('Not valid time range.')
+    timeWhereClause = '({})'.format(' OR '.join(timeWhereClauseArray))
+    return timeWhereClause
+
+def getTimeRangeSumClauseForNeo4j(config, type):
+    def process_quarter(y, m, lastQuarter = 0):
+        q = math.ceil(m / 3)
+        if q != lastQuarter: timeRangeSumClauseArray.append([])
+        timeRangeSumClauseArray[len(timeRangeSumClauseArray) - 1].append('COALESCE({}_{}{}, 0.0)'.format(type, y, m))
+        lastQuarter = q
+    def process_year(y, m, lastYear = 0):
+        if y != lastYear: timeRangeSumClauseArray.append([])
+        timeRangeSumClauseArray[len(timeRangeSumClauseArray) - 1].append('COALESCE({}_{}{}, 0.0)'.format(type, y, m))
+        lastYear = y
+    timeRangeSumClauseArray = []
+    if config.get('groupTimeRange') == 'month':
+        # for every month individual, every element belongs to a individual element
+        forEveryMonthByConfig(config, lambda y, m: timeRangeSumClauseArray.append(['COALESCE({}_{}{}, 0.0)'.format(type, y, m)]))
+    elif config.get('groupTimeRange') == 'quarter':
+        # for every quarter, need to find out when to push a new element by quarter
+        lastQuarter = 0
+        forEveryMonthByConfig(config, process_quarter)
+    elif config.get('groupTimeRange') == 'year':
+        # for every year, need to find out when to push a new element by the year;
+        lastYear = 0
+        forEveryMonthByConfig(config, process_year)
```

Comment:
The `groupTimeRange` config is not properly handled now, right?

Reply:
I am not quite familiar with Python, but now with all kinds of `groupTimeRange`, the query will always be like by `month`.

Reply:
> The `groupTimeRange` config is not properly handled now, right?

I don't think I quite understand what you mean.

Reply:
For Neo4j query with different `groupTimeRange` should generate different query to get the proper result. If the `groupTimeRange` is `year` then data in a year should be aggregate into a single element in the result.

But right now since in Python, a function can not access an external variable, so lastYear in `process_year` will always be 0 and `lastYear` update will not effect the external variable. So the data are not aggregated by year now.

I fix this by remove `lastYear` and `lastQuarter`, then use `month` to judge when to push a new element. I will update this change back too.

Reply:
> 

I have already fix it, and I will update it later.

Reply:
`nonloacl` will fix this too.

But I am using

``` python
if m in [1, 4, 7, 10] or len(timeRangeSumClauseArray) == 0: timeRangeSumClauseArray.append([])
```

for `process_quarter` and 

``` python
if m == 1 or len(timeRangeSumClauseArray) == 0: timeRangeSumClauseArray.append([])
```

for `process_year`.

I think this will do too.

Reply:
Yes, both of them will work.